sql: improve the handling of subqueries

knz opened the pull request:
This work is the final part of the "SQL leaf data" RFC which can be implemented for the 2.0 release. It also makes a step forward towards plan caching.

Fixes #21218.
Fixes #21300.
Informs #20298.

cc @RaduBerinde.

There are really two commits of note in here:

### sql: reify logical plans into a dedicated struct

Prior to this patch, most of the `sql` package was written under the
assumption that there is nothing more to a logical plan than a
reference to its root `planNode`. The various APIs were passing
`planNode` references to each other or as return values, as if that
was sufficient to "communicate a logical plan" from one place to
another.

Alas, in reality this is not so: a logical plan is really composed of
(at least) the following things:

- the placeholder information (currently in `planner.semaCtx.Placeholders`)
- the various computed properties of the logical plan in `planner.curPlan`
  (including e.g. `hasStar`, its `cteNameEnvironment`, etc.)
- in the future, the set of its sub-query plans (these currently hang
  off the scalar expression trees, but that's clearly not their place.
  They belong to a data structure alongside the root tree reference).

It is incorrect to move a `planNode` reference to the root of a logical
plan from one place to another without also preserving these various
other things.

The code happened to "mostly work" because most of it didn't truly
move the `planNode` root from one place to another -- despite the
agitation of moving that root reference from one point to another,
most of the code happened to always be working on "the current plan"
for the current planner, that is, the plan that was most recently
compiled.

In that story, `makePlan` produces two things: an explicit `planNode`
reference and an implicit set of other things annotated in the
`planner` object. The explicit thing navigates around the code,
although its true meaning cannot be dissociated from the silent things
laying in `planner`. However its true meaning is not lost, for every
where the `planNode` reference is finally "consumed", the silent
things also get picked up from the `planner` for consumption alongside
it.

One should not expect however that having dissociated state didn't
really matter, or was a mere discomfort.

There was at least one bug which I found while working on that
dissociative disorder: the function `makeInternalPlan()`, meant for
use "within a plan" to compute a sub-result, was documented to
"clobber its placeholder information". Alas, this is not the only
thing it was clobbering: by virtue of calling `makePlan`, it was also
clobbering *everything* associated to the "current plan" -- all the
things computed for the sub-plan, in fact, including the placeholder
information but also the other computed properties. Fortunately, I was
unable to trace a code path that would suggest this bug exhibits
itself in the wild, but it was very much alive nonetheless.

So, what happens in this patch is that all the state associated with a
logical plan is given a name: `planTop`. It is now a struct defined
in `plan.go`.

The implicit state captured in the `planner` is not implicit any more:
the `planner`'s `curPlan` member is now an instance of `planTop`.
Also, the root `planNode` reference is now also stored in `planTop`.

Consequently:

- `makePlan` is now defined to compile the logical plan in
  `planner.curPlan`. It does not "return a logical plan" any more; it
  *prepares the logical plan in the area of the `planner struct`
  (`curPlan`) dedicated to hold the logical plan details.

  `makePlan` is now also clearly documented as clobbering the entire
  `planner.curPlan` field.

- the aforementioned bug is fixed by ensuring that the places
  where "the current logical plan state must be saved" save the entire
  `planTop` struct, not just the placeholder information.

- the ritualistic, but mistakenly dissociated, passing around of
  `planNode` root references is removed throughout most of the code,
  in favor of clearer and more deliberate naming of the "current
  logical plan" via explicit accesses to `p.curPlan`.

There are two APIs that could have been modified accordingly, but were
not in this patch, considering a potential future optimization:
`execDistSQL` and `execClassic`. These could also operate "on the
current logical plan" and pick up the tree root from `planner.curPlan`
like the rest. However I chose not to do this in this patch, because I
foresee that these methods could be called on individual sub-query
plans in a later patch.

### sql, sem/tree: avoid storing sub-query plans inside scalar expressions 

Prior to this patch, logical planning for (uncorrelated) subqueries
would work more-or-less like this:

- a first recursion would traverse all scalar expressions (`tree.Expr`) and
  replace any `tree.Subquery` by a `sql.subquery` object.
  It would also recursively create a logical plan and store it inside
  the `sql.subquery` object for later use.
- during logical planning, another recursion would traverse all scalar
  expressions and perform logical planning optimizations in turn
  on each found `sql.subquery` object.
- finally just before execution, another recursion would traverse  all
  scalar expressions (again!) to start the sub-query plans and compute
  their (datum) results.

This patch changes this according to the RFC on "SQL leaf data"
(`docs/RFCS/20170926_sql_aux_data.md`): the initial traversal now
*extracts* the subquery details into a top-level slice maintained for
the entire logical plan. Further processing that needs to "operate on
all subqueries" can find the sub-query plans, if any, using the
top-level slice. No further traversal of all scalar expressions is
needed.

Salient changes:

- the slice of sub-query objects (`sql.subquery`) is stored in the new
  `planTop` struct recently introduced. That object is responsible for
  closing the sub-query plans when its own `close()` method is called.

- because recursion to all expressions is not needed any more to
  collect subquery plan, the code in `walk.go` is greatly simplified.

- the `tree.Subquery` node is extended so that it can remain in the
  scalar expression tree up to final evaluation. To achieve this, the
  index into the top-level slice, as well as the computed `types.T`,
  is now stored in the `tree.Subquery`. Its `Eval` method, which now
  needs to be functional, retrieves the computed subquery datum using
  a new `EvalSubquery` method on the `evalCtx.Planner` object.

- the code for EXPLAIN(PLAN) needs to be able to observe sub-queries
  locally without running them, so now it needs to prevent the
  sub-queries of the plan being observed from being stored in the
  top-level slice (previously, it didn't need to prevent anything --
  its sub-query plans were simply those present in scalar expression
  trees of its observed plan). It achieves this by temporarily
  overriding the sub-query slice while the explainPlanNode is being
  built.

- the code for EXPLAIN is extended to show the sub-queries at the top
  of the logical plan by introducing a new virtual "root" node when
  there are subqueries present.

  In addition, the subquery execution mode is now also displayed in
  the EXPLAIN output, to aid troubleshooting.

  Also, the link/relationship between the
  `tree.Subquery` nodes in scalar expressions and the top-level slice
  of subquery is now revealed in the output of EXPLAIN by using
  symbolic identiifers for the subqueries.

  For example:

```
root@:26257/> explain(exprs) select 123+(select 1);
+-------------------------+-----------+-------------+
|          Tree           |   Field   | Description |
+-------------------------+-----------+-------------+
| root                    |           |             |
|  ├── render             |           |             |
|  │    │                 | render 0  | 123 + @S1   |
|  │    └── emptyrow      |           |             |
|  └── subquery           |           |             |
|       │                 | id        | @S1         |
|       │                 | sql       | (SELECT 1)  |
|       │                 | exec mode | one row     |
|       └── render        |           |             |
|            │            | render 0  |           1 |
|            └── emptyrow |           |             |
+-------------------------+-----------+-------------+

root@:26257/> explain(exprs) select 123+(select 1 from (select 456+(select 2) from t.kv));
+--------------------------+-----------+-----------------------------------------------------+
|           Tree           |   Field   |                     Description                     |
+--------------------------+-----------+-----------------------------------------------------+
| root                     |           |                                                     |
|  ├── render              |           |                                                     |
|  │    │                  | render 0  | 123 + @S2                                           |
|  │    └── emptyrow       |           |                                                     |
|  ├── subquery            |           |                                                     |
|  │    │                  | id        | @S1                                                 |
|  │    │                  | sql       | (SELECT 2)                                          |
|  │    │                  | exec mode | one row                                             |
|  │    └── render         |           |                                                     |
|  │         │             | render 0  |                                                   2 |
|  │         └── emptyrow  |           |                                                     |
|  └── subquery            |           |                                                     |
|       │                  | id        | @S2                                                 |
|       │                  | sql       | (SELECT 1 FROM (SELECT 456 + (SELECT 2) FROM t.kv)) |
|       │                  | exec mode | one row                                             |
|       └── render         |           |                                                     |
|            │             | render 0  |                                                   1 |
|            └── render    |           |                                                     |
|                 │        | render 0  | NULL                                                |
|                 └── scan |           |                                                     |
|                          | table     | kv@primary                                          |
|                          | spans     | ALL                                                 |
+--------------------------+-----------+-----------------------------------------------------+
```

knz committed:
sql/tree: remove ExistsExpr

It is really a flag in the Subquery node. This simplifies the code
generally.

Release note: none.

knz committed:
sql: make EXPLAIN's option NONORMALIZE apply only to sub-plan

Prior to this pathc, using EXPLAIN(NONORMALIZE) in any sub-query would cause
expression normalization to be disabled for any subsequent EXPLAIN sub-query,
not just this one.

This patch ensures that the option NONORMALIZE only applies to the
query explained by the EXPLAIN where the option appears.

Release note: none.

knz committed:
sql: fix a misleading comment in explain.go

Release note: none

knz committed:
sql: make the default value of subqueryExecMode invalid

... so as to facilitate troubleshooting of insufficient initializations.

Release note: none

knz committed:
sem/tree: make (*Subquery).Walk agnostic of struct fields

So that we don't have to care about the Walk method when adding flags
to the struct.

Release note: none

knz committed:
sem/tree: add a space between ARRAY and its subquery

Release note: none

knz committed:
sql: reify logical plans into a dedicated struct

Prior to this patch, most of the `sql` package was written under the
assumption that there is nothing more to a logical plan than a
reference to its root `planNode`. The various APIs were passing
`planNode` references to each other or as return values, as if that
was sufficient to "communicate a logical plan" from one place to
another.

Alas, in reality this is not so: a logical plan is really composed of
(at least) the following things:

- the placeholder information (currently in `planner.semaCtx.Placeholders`)
- the various computed properties of the logical plan in `planner.curPlan`
  (including e.g. `hasStar`, its `cteNameEnvironment`, etc.)
- in the future, the set of its sub-query plans (these currently hang
  off the scalar expression trees, but that's clearly not their place.
  They belong to a data structure alongside the root tree reference).

It is incorrect to move a `planNode` reference to the root of a logical
plan from one place to another without also preserving these various
other things.

The code happened to "mostly work" because most of it didn't truly
move the `planNode` root from one place to another -- despite the
agitation of moving that root reference from one point to another,
most of the code happened to always be working on "the current plan"
for the current planner, that is, the plan that was most recently
compiled.

In that story, `makePlan` produces two things: an explicit `planNode`
reference and an implicit set of other things annotated in the
`planner` object. The explicit thing navigates around the code,
although its true meaning cannot be dissociated from the silent things
laying in `planner`. However its true meaning is not lost, for every
where the `planNode` reference is finally "consumed", the silent
things also get picked up from the `planner` for consumption alongside
it.

One should not expect however that having dissociated state didn't
really matter, or was a mere discomfort.

There was at least one bug which I found while working on that
dissociative disorder: the function `makeInternalPlan()`, meant for
use "within a plan" to compute a sub-result, was documented to
"clobber its placeholder information". Alas, this is not the only
thing it was clobbering: by virtue of calling `makePlan`, it was also
clobbering *everything* associated to the "current plan" -- all the
things computed for the sub-plan, in fact, including the placeholder
information but also the other computed properties. Fortunately, I was
unable to trace a code path that would suggest this bug exhibits
itself in the wild, but it was very much alive nonetheless.

So, what happens in this patch is that all the state associated with a
logical plan is given a name: `planTop`. It is now a struct defined
in `plan.go`.

The implicit state captured in the `planner` is not implicit any more:
the `planner`'s `curPlan` member is now an instance of `planTop`.
Also, the root `planNode` reference is now also stored in `planTop`.

Consequently:

- `makePlan` is now defined to compile the logical plan in
  `planner.curPlan`. It does not "return a logical plan" any more; it
  *prepares the logical plan in the area of the `planner struct`
  (`curPlan`) dedicated to hold the logical plan details.

  `makePlan` is now also clearly documented as clobbering the entire
  `planner.curPlan` field.

- the aforementioned bug is fixed by ensuring that the places
  where "the current logical plan state must be saved" save the entire
  `planTop` struct, not just the placeholder information.

- the ritualistic, but mistakenly dissociated, passing around of
  `planNode` root references is removed throughout most of the code,
  in favor of clearer and more deliberate naming of the "current
  logical plan" via explicit accesses to `p.curPlan`.

There are two APIs that could have been modified accordingly, but were
not in this patch, considering a potential future optimization:
`execDistSQL` and `execClassic`. These could also operate "on the
current logical plan" and pick up the tree root from `planner.curPlan`
like the rest. However I chose not to do this in this patch, because I
foresee that these methods could be called on individual sub-query
plans in a later patch.

Release note: None

knz committed:
sql: rename 'classic exec' -> 'local exec'

Release note: none

knz committed:
sql: avoid calling checkResultTypes on every result row

Prior to this patch, the code for local execution would unnecessarily
call `checkResultTypes()` on each row being processed. This is
unnecessary because every row has the same type as the declared column
plans.

This patch reduces this overhead by checking the types just once.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21305)
<!-- Reviewable:end -->

knz commented:
@RaduBerinde now that we have all the SQ plans at the top level I propose to select distributed vs local execution for each of them in turn, computing the result datums in this way, and then apply again the dist/local choice on the main plan after all the SQ datums are available. In this way the distsql phy planner doesn't need to reject subqueries any more -- it can simply serialize them using the precomputed datum as representation. What do you think?

petermattis commented:
Basing this on top of #21234 makes it almost unreviewable. I wonder how reviewable will handle the rebase. You may need to close this PR and reopen it. Maybe. I'm going to hold off on reviewing until #21234 is merged.

---

Review status: 0 of 141 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-:-L2GqPbSxdZTB6mgQ0Om:b37xsvp)*
<!-- Sent from Reviewable.io -->

knz commented:
I'm not sure about you but Reviewable currently plainly refuses to open the PR properly (it collapses all commits into one).  Indeed the review may need to wait until #21234 merges. Although this could perhaps happen soon? I reckon there is not much left preventing it?

petermattis commented:
Yeah, reviewable is only showing me a single change. I think you'll need to close and reopen this. 

---

Review status: 0 of 141 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-:-L2Gs_euOu0T1qr9yyT7:b-2k9y66)*
<!-- Sent from Reviewable.io -->

knz committed:
sql, sem/tree: avoid storing sub-query plans inside scalar expressions

Prior to this patch, logical planning for (uncorrelated) subqueries
would work more-or-less like this:

- a first recursion would traverse all scalar expressions (`tree.Expr`) and
  replace any `tree.Subquery` by a `sql.subquery` object.
  It would also recursively create a logical plan and store it inside
  the `sql.subquery` object for later use.
- during logical planning, another recursion would traverse all scalar
  expressions and perform logical planning optimizations in turn
  on each found `sql.subquery` object.
- finally just before execution, another recursion would traverse  all
  scalar expressions (again!) to start the sub-query plans and compute
  their (datum) results.

This patch changes this according to the RFC on "SQL leaf data"
(`docs/RFCS/20170926_sql_aux_data.md`): the initial traversal now
*extracts* the subquery details into a top-level slice maintained for
the entire logical plan. Further processing that needs to "operate on
all subqueries" can find the sub-query plans, if any, using the
top-level slice. No further traversal of all scalar expressions is
needed.

Salient changes:

- the slice of sub-query objects (`sql.subquery`) is stored in the new
  `planTop` struct recently introduced. That object is responsible for
  closing the sub-query plans when its own `close()` method is called.

- because recursion to all expressions is not needed any more to
  collect subquery plan, the code in `walk.go` is greatly simplified.

- the `tree.Subquery` node is extended so that it can remain in the
  scalar expression tree up to final evaluation. To achieve this, the
  index into the top-level slice, as well as the computed `types.T`,
  is now stored in the `tree.Subquery`. Its `Eval` method, which now
  needs to be functional, retrieves the computed subquery datum using
  a new `EvalSubquery` method on the `evalCtx.Planner` object.

- the code for EXPLAIN(PLAN) needs to be able to observe sub-queries
  locally without running them, so now it needs to prevent the
  sub-queries of the plan being observed from being stored in the
  top-level slice (previously, it didn't need to prevent anything --
  its sub-query plans were simply those present in scalar expression
  trees of its observed plan). It achieves this by temporarily
  overriding the sub-query slice while the explainPlanNode is being
  built.

- the code for EXPLAIN is extended to show the sub-queries at the top
  of the logical plan by introducing a new virtual "root" node when
  there are subqueries present.

  In addition, the subquery execution mode is now also displayed in
  the EXPLAIN output, to aid troubleshooting.

  Also, the link/relationship between the
  `tree.Subquery` nodes in scalar expressions and the top-level slice
  of subquery is now revealed in the output of EXPLAIN by using
  symbolic identiifers for the subqueries.

  For example:

  ```
root@:26257/> explain(exprs) select 123+(select 1);
+-------------------------+-----------+-------------+
|          Tree           |   Field   | Description |
+-------------------------+-----------+-------------+
| root                    |           |             |
|  ├── render             |           |             |
|  │    │                 | render 0  | 123 + @S1   |
|  │    └── emptyrow      |           |             |
|  └── subquery           |           |             |
|       │                 | id        | @S1         |
|       │                 | sql       | (SELECT 1)  |
|       │                 | exec mode | one row     |
|       └── render        |           |             |
|            │            | render 0  |           1 |
|            └── emptyrow |           |             |
+-------------------------+-----------+-------------+

root@:26257/> explain(exprs) select 123+(select 1 from (select 456+(select 2) from t.kv));
+--------------------------+-----------+-----------------------------------------------------+
|           Tree           |   Field   |                     Description                     |
+--------------------------+-----------+-----------------------------------------------------+
| root                     |           |                                                     |
|  ├── render              |           |                                                     |
|  │    │                  | render 0  | 123 + @S2                                           |
|  │    └── emptyrow       |           |                                                     |
|  ├── subquery            |           |                                                     |
|  │    │                  | id        | @S1                                                 |
|  │    │                  | sql       | (SELECT 2)                                          |
|  │    │                  | exec mode | one row                                             |
|  │    └── render         |           |                                                     |
|  │         │             | render 0  |                                                   2 |
|  │         └── emptyrow  |           |                                                     |
|  └── subquery            |           |                                                     |
|       │                  | id        | @S2                                                 |
|       │                  | sql       | (SELECT 1 FROM (SELECT 456 + (SELECT 2) FROM t.kv)) |
|       │                  | exec mode | one row                                             |
|       └── render         |           |                                                     |
|            │             | render 0  |                                                   1 |
|            └── render    |           |                                                     |
|                 │        | render 0  | NULL                                                |
|                 └── scan |           |                                                     |
|                          | table     | kv@primary                                          |
|                          | spans     | ALL                                                 |
+--------------------------+-----------+-----------------------------------------------------+
  ```

Release note (sql change): the output of EXPLAIN is enhanced when
applied to statements containing scalar sub-queries.

knz commented:
Reviewable is un-broken after the rebase.

petermattis commented:
Using `tree.Subquery.Idx` vs replacing the `tree.Subquery` node with `subquery` seems equivalent if I squint. With either approach we can keep a slice of the `subquery` nodes found during `replaceSubqueries` which seems to be the big benefit to this change. I don't particularly mind that part of the change, just pointing out that it is shuffling around things a bit more than necessary. 

---

Review status: 0 of 29 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/sql/plan.go, line 222 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HEwRhsIk3NObbNduP:-L2HEwRhsIk3NObbNduQ:b-nvs7nj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/plan.go#L222)):*
> ```Go
> 
> 	// sqPlans contains all the sub-query plans.
> 	sqPlans []subquery
> ```

My brain wants to interpret this name as `square plans`. Can we name this `subqueryPlans` or just `subqueries`?

---

*[pkg/sql/subquery.go, line 363 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HENOXsDWExbXDdwfq:-L2HENOXsDWExbXDdwfr:b-eolmod) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/subquery.go#L363)):*
> ```Go
> }
> 
> func (v *subqueryVisitor) replaceSubquery(
> ```

You're not really replacing the subquery anymore, so this is now misnamed. 

---

*[pkg/sql/subquery.go, line 388 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HEBHOFUlLD4xKjIyy:-L2HEBHOFUlLD4xKjIyz:b-fkmr6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/subquery.go#L388)):*
> ```Go
> 
> 	v.curPlan.sqPlans = append(v.curPlan.sqPlans, subquery{subquery: sub, plan: plan})
> 	sub.Idx = len(v.curPlan.sqPlans) // Note: subquery node are 1-indexed, so that 0 remains invalid.
> ```

I think this comment would be better placed (or perhaps duplicated) on `tree.Subquery.Idx`.

---

*[pkg/sql/logictest/testdata/logic_test/subquery, line 170 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HDYU5EtGx6v61PAfs:-L2HDYU5EtGx6v61PAft:b2j1r87) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/logictest/testdata/logic_test/subquery#L170)):*
> ```
>       │                       id         @S1
>       │                       sql        (SELECT 42)
>       │                       exec mode  one row
> ```

I'm not clear on why we would prefer displaying the subqueries all connected to the root vs the preview display where there is a single logical tree with the subqueries display where they occur. Seems like the display proposed in this PR is highlighting an implementation detail. Couldn't you still display the subquery in-situ? See my suggestion on `tree.Subquery.Format`.

---

*[pkg/sql/sem/tree/expr.go, line 789 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HCwNptB3fw4VSDWDP:-L2HCwNptB3fw4VSDWDQ:btjmbfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/sem/tree/expr.go#L789)):*
> ```Go
> type Subquery struct {
> 	Select SelectStatement
> 	Exists bool
> ```

Well that certainly simplifies the walk to find subqueries. Obvious in hindsight. I wish I had thought of it. 

---

*[pkg/sql/sem/tree/expr.go, line 792 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HBjijnY0LxtUyq8w2:-L2HBjijnY0LxtUyq8w3:brw4x2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/sem/tree/expr.go#L792)):*
> ```Go
> 
> 	// Idx indicates the known number of the sub-query in the top-level
> 	// plan.
> ```

I find the `known number` phrase a bit confusing. Perhaps `Idx is a query unique index for the subquery`.

---

*[pkg/sql/sem/tree/expr.go, line 796 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HBHfniOCkKofBvFa8:-L2HBHfniOCkKofBvFa9:b-7x9pt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/sem/tree/expr.go#L796)):*
> ```Go
> 
> 	// Typ is exposed to be populated by sql.replaceSubqueries().
> 	Typ types.T
> ```

Should this be `typAnnotation`. In general, we seem to not export the `types.T` field for other `TypedExpr`, but use `ResolvedType()`.

---

*[pkg/sql/sem/tree/expr.go, line 817 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HCI9YeDmLzqIgN5KR:-L2HCI9ZbBc3WTIEBvu9:b-qgejv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/sem/tree/expr.go#L817)):*
> ```Go
> func (node *Subquery) Format(ctx *FmtCtx) {
> 	if ctx.HasFlags(FmtSymbolicSubqueries) {
> 		ctx.Printf("@S%d", node.Idx)
> ```

We could still format the actual subquery here if you added `FmtCtx.subqueryFormat`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-:-L2Gxch89kNzttfRl8uj:b-u3n864)*
<!-- Sent from Reviewable.io -->

knz commented:
> Using tree.Subquery.Idx vs replacing the tree.Subquery node with subquery seems equivalent if I squint.

Well not so much -- replacing the subquery nodes by `*sql.subquery` prevents true AST reuse. With the previous code there would be an AST rewrite (via `tree.WalkExpr`) each time a query was executed after prepare, with this new code this AST churn disappears.

---

Review status: 0 of 29 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/plan.go, line 222 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HEwRhsIk3NObbNduP:-L2HHyUvjbsEDwSbdAw6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/plan.go#L222)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

My brain wants to interpret this name as `square plans`. Can we name this `subqueryPlans` or just `subqueries`?
</blockquote></details>

Done.

---

*[pkg/sql/subquery.go, line 388 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HEBHOFUlLD4xKjIyy:-L2HJBcwSddsO_QYGfF2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/subquery.go#L388)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think this comment would be better placed (or perhaps duplicated) on `tree.Subquery.Idx`.
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/subquery, line 170 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HDYU5EtGx6v61PAfs:-L2HJrWga2iq0Dr0rq2y:bkk7dji) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/logictest/testdata/logic_test/subquery#L170)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm not clear on why we would prefer displaying the subqueries all connected to the root vs the preview display where there is a single logical tree with the subqueries display where they occur. Seems like the display proposed in this PR is highlighting an implementation detail. Couldn't you still display the subquery in-situ? See my suggestion on `tree.Subquery.Format`.
</blockquote></details>

So there are two arguments to be made here:

1) this code is simpler and lighter. This way the FmtCtx struct does not need to be extended, and we don't one more of these callbacks into the `sql` package from the node formatter recursion (IMHO we should have no such callback whatsoever). Look at it the other way around: if this new layout was the one we had from the beginning, would the increase in complexity required to show subqueries "next to where they are used" worth the improved UX?

2) I actually prefer this layout better. It highlights that subqueries are treated as variables in scalar expressions.

---

*[pkg/sql/sem/tree/expr.go, line 792 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HBjijnY0LxtUyq8w2:-L2HJTdgxU83mgN18w3R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/sem/tree/expr.go#L792)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I find the `known number` phrase a bit confusing. Perhaps `Idx is a query unique index for the subquery`.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/expr.go, line 796 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HBHfniOCkKofBvFa8:-L2HIk8qg3oBjRCVpl4W:b-xu82yc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/sem/tree/expr.go#L796)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should this be `typAnnotation`. In general, we seem to not export the `types.T` field for other `TypedExpr`, but use `ResolvedType()`.
</blockquote></details>

Ok I'll take the code sharing. Done.

---

*[pkg/sql/sem/tree/expr.go, line 817 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HCI9YeDmLzqIgN5KR:-L2HKTDCTMFSwm35DeIV:b-apayz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/sem/tree/expr.go#L817)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We could still format the actual subquery here if you added `FmtCtx.subqueryFormat`. 
</blockquote></details>

See my comment above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-:-L2HKos5t6MIux3cV1fU:bfzrxsn)*
<!-- Sent from Reviewable.io -->

knz commented:
Note: this PR breaks two sets of CCL tests, and the test logs point specifically to `testutils/jobutils/jobs_verification.go`. I suspect there's an error there since every other test succeeds.  @dt can I pick your brain as to what this code is meant to do? I am having a hard time reverse-engineering it.

knz commented:
Review status: 0 of 29 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/subquery.go, line 363 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HENOXsDWExbXDdwfq:-L2HMGMNSX-gQEzO-mtG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/subquery.go#L363)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You're not really replacing the subquery anymore, so this is now misnamed. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> Well not so much -- replacing the subquery nodes by *sql.subquery prevents true AST reuse. With the previous code there would be an AST rewrite (via tree.WalkExpr) each time a query was executed after prepare, with this new code this AST churn disappears.

Doesn't the replacement of `*tree.Subquery` with `*sql.subquery` happen during prepare? I think I'm missing something obvious here.

---

Review status: 0 of 29 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/logictest/testdata/logic_test/subquery, line 170 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2HDYU5EtGx6v61PAfs:-L2HNrW-V0fE-1LcI3Le:b-nbn3m1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c03e1761208c391244fa0c6ab5fe2ec404cdd347/pkg/sql/logictest/testdata/logic_test/subquery#L170)):*
> Look at it the other way around: if this new layout was the one we had from the beginning, would the increase in complexity required to show subqueries "next to where they are used" worth the improved UX?

Probably.

> I actually prefer this layout better. It highlights that subqueries are treated as variables in scalar expressions.

Ok, we have 2 differing opinions. I'm curious to know what others think. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-:-L2HO5Q2xzZfozuekD5a:b-4v7xc)*
<!-- Sent from Reviewable.io -->

knz commented:
> Doesn't the replacement of *tree.Subquery with *sql.subquery happen during prepare?

No it happens during logical planning, which we currently (until plan caching becomes a thing) re-do on every execute.

For plan caching the analysis must be split into two different phases:
- the extraction of subqueries ASTs into the top plan must happen once during parse (or soon afterwards)
- the compilation of the logical plan for subqueries during prepare

> I think I'm missing something obvious here.

It's not obvious...

---

Review status: 0 of 29 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-:-L2K3EBRaVANneEBJPdd:b-iaud6)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks for the great commit messages, they help a lot. Very nice changes!

The EXPLAIN plans are much more readable, the `subquery` node helps a lot. I kind of agree with Peter that it would be nice if the `subquery` nodes were hanging off of the node which uses them, but I don't think it's a big deal (so IMO not worth doing if it adds complexity; and one can argue that it's more accurate to hang them off of the root since that's effectively how we are running them).

petermattis commented:
> No it happens during logical planning, which we currently (until plan caching becomes a thing) re-do on every execute.

I thought we reused the parsed and type-checked AST during execute. Is that not true? 

> For plan caching the analysis must be split into two different phases:
> 
> the extraction of subqueries ASTs into the top plan must happen once during parse (or soon afterwards)
> the compilation of the logical plan for subqueries during prepare

Got it. I agree that we need to extract subqueries once and not every time a query is executed. My point has been that there isn't much difference between setting `tree.Subquery.Idx` vs replacing `*tree.Subquery` with some sort of `*subquery` node (not necessarily the existing `*sql.subquery`). Either way, we'd want a slice of subquery plans that we can execute. Either way, you're munging the AST after parsing.

---

Review status: 0 of 29 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-:-L2Lh0ypxFM8IMQx48Zq:b-wbqugh)*
<!-- Sent from Reviewable.io -->

knz commented:
> I thought we reused the parsed and type-checked AST during execute.

Parsed but not type-checked.

You forget one step in-between: name resolution. That needs the db schema, and so we need to re do this at plan time because we don't trash the AST cache on schema changes.

knz commented:
Also:

> Either way, you're munging the AST after parsing.

Just once though, if there are multiple executes for a single prepare.

We only rewrite the AST during walk if the old/new nodes compare equal structurally. After the first iteration, the Idx field gets rewritten to the same value by the recursion, so the AST doesn't need to go through GC any more. It's a fairly subtle distinction that likely does not matter wrt performance, just wanted to clarify.

petermattis commented:
> Parsed but not type-checked.
>
> You forget one step in-between: name resolution. That needs the db schema, and so we need to re do this at plan time because we don't trash the AST cache on schema changes.

I wasn't forgetting about name resolution. We definitely type check at prepare. Are we not keeping the type checked query around? We certainly will have to with plan caching. 

> We only rewrite the AST during walk if the old/new nodes compare equal structurally. After the first iteration, the Idx field gets rewritten to the same value by the recursion, so the AST doesn't need to go through GC any more. It's a fairly subtle distinction that likely does not matter wrt performance, just wanted to clarify.

This is exactly my point. If we replaced `*tree.Subquery` with `*subquery` during prepare, a subsequent execute wouldn't have to make any further changes to the AST. Now this would be with a different version of `*subquery` that didn't contain any execution state, just the plan. I feel we're talking past each other.

knz commented:
> Are we not keeping the type checked query around?

No

> We certainly will have to with plan caching.

yes

> If we replaced *tree.Subquery with *subquery during prepare, a subsequent execute wouldn't have to make any further changes to the AST. 

Yes agreed. 

I was merely pointing out that this PR in its current state does reduce allocation churn, albeit not by much.

knz commented:
After a head scratching troubleshooting session with Andrei, we found out that this PR is currently failing CI because I broke limit propagation in the last commit. 

The debugging process revealed that our EXPLAIN/planToString output is incomplete, which I'll address separately, but I think the remaining problem with this PR must be a trivial oversight. Still looking for it though.

knz commented:
Found it - I had mistakenly inverted two lines in `startPlan`. Fingers crossed, the change should be complete/correct now.

knz commented:
yep it's green now

petermattis commented:
> I was merely pointing out that this PR in its current state does reduce allocation churn, albeit not by much.

Roger. I don't think we need to be concerned about allocations at this level. Regardless, the approach of using `tree.Subquery.Idx` is fine with me.

Someone else in SQL-land should give this PR a more thorough review than I can give.

knz commented:
Review status: 0 of 30 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/executor_opt_interface.go, line 83 at r20](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2pEd9gQsoW1lXHmZ3j:-L2pEd9gQsoW1lXHmZ3k:b3eobum) ([raw file](https://github.com/cockroachdb/cockroach/blob/8625bb4b9c6d3e74caaf0f3446bb90ca1ed8bea6/pkg/sql/executor_opt_interface.go#L83)):*
> ```Go
> 		qualifyNames: true,
> 	}
> 	explainNode, err := en.planner.makeExplainPlanNode(
> ```

@RaduBerinde your change broke my PR and I do not know how to fix this.
The last parameter of `makeExplainPlanNode` must be the original statement AST, or the function must be modified somehow so that it properly captures the collection of subquery nodes. Please amend my commit in the way you see fit. I am building a series of changes upon this PR and I would prefer not to have to context switch to look at this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Ill take a look. I remember removing an unused arg, we can surely add it back.

RaduBerinde commented:
Review status: 0 of 30 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor_opt_interface.go, line 83 at r20](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L2pEd9gQsoW1lXHmZ3j:-L2qaufBxrmzzVy0dtkY:bcw13qp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8625bb4b9c6d3e74caaf0f3446bb90ca1ed8bea6/pkg/sql/executor_opt_interface.go#L83)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

@RaduBerinde your change broke my PR and I do not know how to fix this.
The last parameter of `makeExplainPlanNode` must be the original statement AST, or the function must be modified somehow so that it properly captures the collection of subquery nodes. Please amend my commit in the way you see fit. I am building a series of changes upon this PR and I would prefer not to have to context switch to look at this.
</blockquote></details>

The diff below seems to do the trick. Note that we don't have to worry about subqueries in this context, the plan we generate through this interface will never contain one.

```diff
diff --git a/pkg/sql/executor_opt_interface.go b/pkg/sql/executor_opt_interface.go
index f9103cbf8..52b219a5e 100644
--- a/pkg/sql/executor_opt_interface.go
+++ b/pkg/sql/executor_opt_interface.go
@@ -80,17 +80,18 @@ func (en *execNode) Explain() ([]tree.Datums, error) {
 		showExprs:    true,
 		qualifyNames: true,
 	}
-	explainNode, err := en.planner.makeExplainPlanNode(
+	explainNode, err := en.planner.makeExplainPlanNodeWithPlan(
 		context.TODO(), flags, false /* expanded */, false /* optimized */, en.plan,
 	)
 	if err != nil {
 		return nil, err
 	}
-	explainNode := execNode{
+	// Run the node.
+	explainExecNode := execNode{
 		execFactory: en.execFactory,
 		plan:        explainNode,
 	}
-	return explainNode.Run()
+	return explainExecNode.Run()
 }
 
 // Run is part of the opt.ExecNode interface.
diff --git a/pkg/sql/explain_plan.go b/pkg/sql/explain_plan.go
index 5ef44ff8a..993c08040 100644
--- a/pkg/sql/explain_plan.go
+++ b/pkg/sql/explain_plan.go
@@ -84,7 +84,14 @@ func (p *planner) makeExplainPlanNode(
 	if err != nil {
 		return nil, err
 	}
+	return p.makeExplainPlanNodeWithPlan(ctx, flags, expanded, optimized, plan)
+}
 
+// newExplainPlanNodeWithPlan instantiates a planNode that EXPLAINs an
+// underlying plan.
+func (p *planner) makeExplainPlanNodeWithPlan(
+	ctx context.Context, flags explainFlags, expanded, optimized bool, plan planNode,
+) (planNode, error) {
 	columns := explainPlanColumns
 
 	if flags.showMetadata {
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
> @RaduBerinde now that we have all the SQ plans at the top level I propose to select distributed vs local execution for each of them in turn, computing the result datums in this way, and then apply again the dist/local choice on the main plan after all the SQ datums are available. In this way the distsql phy planner doesn't need to reject subqueries any more -- it can simply serialize them using the precomputed datum as representation. What do you think?

Sorry, I missed this earlier. Yes, what you suggest would be great.

knz commented:
omg it's green again

knz commented:
welp, talked too soon. Another conflict.

knz commented:
yep green again

jordanlewis commented:
LGTM!

I have a couple of comments, but I don't want them to block merging this PR. We should debate and resolve them later.

1. You identified a major dissociation as you put it between the fact that all of the plan creation methods return a plan node, but plan nodes themselves aren't a full encapsulation of a plan. You solved that by making these methods mutate the planner instead. I think that's a little unpleasant - I'd rather see the world move to a place where (completely specified) planTops are passed around and not automatically mutated.
2. I love the concept behind introducing a root node for explain on queries with subqueries - it spells out to the user that subqueries are executed separately. But I wonder if it will be difficult to consistently parse for users who want to write tools around our explain output, or just to explain to users in general despite the fact that it seems like it would improve things. Another option is to leave the subquery inside the normal hierarchy, but explain via some modifier that it will just get executed once. I see now that there's been some debate about this already...

knz commented:
Thanks Jordan for the review!

Regarding point 1: I agree. However passing around of a planTop was more work than strictly needed for the proper handling of subqueries, that's why I didn't investigate that further in this PR. Note that handling planTop like you suggest would have required the commit in the PR as a starting point anyway.

Regarding point 2: all the subqueries are currently executed just once. The location of the use of that subquery's result(s) is visible in the output of `explain(exprs)` with the syntax `@Sxxxx`. I am against investing too much intelligence into polishing the UX for EXPLAIN because the entire semantics of this will be overhauled in the next release anyway (due to new planning code).

What do you think of my reaction? How would you like me to follow up?

jordanlewis commented:
1. Makes sense to me. I wouldn't suggest that you carry that change out fully in this PR. I just wanted to note my reaction to the change. Let's do it later!
2. Fair point about the upcoming overhaul.

Ship it!!

knz commented:
Yay :tada:

andreimatei commented:
Review status: 0 of 30 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parallel_stmts.go, line 83 at r30](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L3ym9P0PRY9f44p_R6x:-L3ym9P0PRY9f44p_R6y:bihotbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e140bedd72924c16164d0ffa6a15ee2a3632c5/pkg/sql/parallel_stmts.go#L83)):*
> ```Go
> // details.
> func (pq *ParallelizeQueue) Add(params runParams, exec func() error) error {
> 	// plan is used as a hash key in the data structures, we don't
> ```

references to a "hash key" both here and in the new comment below left me scratching my head for a while. First of all, I don't know what this comment is trying to tell me... That we don't use this `plan` variable? Then why do we even declare it?
Second of all, the fact that go's maps are hashtables doesn't matter for anything. I'd strike this comment (and hopefully the variable can go too).

---

*[pkg/sql/parallel_stmts.go, line 86 at r30](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L3yl2MV53kpzWIYygQO:-L3yl2MV53kpzWIYygQP:b-21dr32) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e140bedd72924c16164d0ffa6a15ee2a3632c5/pkg/sql/parallel_stmts.go#L86)):*
> ```Go
> 	// consider it further here.
> 	plan := params.p.curPlan.plan
> 	prereqs, finishLocked, err := pq.insertInQueue(params)
> ```

I think the `insertInQueue` interface changed in the wrong direction - why is it taking a `runParams` instead of just a `plan`? I couldn't see a reason.
It was always bad, but I think it became worse now - at least before it seemed likely to a reader that the second argument (the plan) matters and the first one is random crap. 

---

*[pkg/sql/parallel_stmts.go, line 88 at r30](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L3yl-CNFC21ARnFyVIH:-L3yl-CNFC21ARnFyVII:btyzcpg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e140bedd72924c16164d0ffa6a15ee2a3632c5/pkg/sql/parallel_stmts.go#L88)):*
> ```Go
> 	prereqs, finishLocked, err := pq.insertInQueue(params)
> 	if err != nil {
> 		plan.Close(params.ctx)
> ```

should this be `params.p.curPlan.close()` ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 30 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parallel_stmts.go, line 83 at r30](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L3ym9P0PRY9f44p_R6x:-L4c6O9xMTPfqxie6qck:b-lqrp2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e140bedd72924c16164d0ffa6a15ee2a3632c5/pkg/sql/parallel_stmts.go#L83)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

references to a "hash key" both here and in the new comment below left me scratching my head for a while. First of all, I don't know what this comment is trying to tell me... That we don't use this `plan` variable? Then why do we even declare it?
Second of all, the fact that go's maps are hashtables doesn't matter for anything. I'd strike this comment (and hopefully the variable can go too).
</blockquote></details>

the point being that the /plan/ (i.e. the tree data structure) is not used; the only thing that's used is the value of the address to the top node, used as hash key.

---

*[pkg/sql/parallel_stmts.go, line 86 at r30](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L3yl2MV53kpzWIYygQO:-L4c6_-US9f5WPibwIsS:b-yt0tnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e140bedd72924c16164d0ffa6a15ee2a3632c5/pkg/sql/parallel_stmts.go#L86)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think the `insertInQueue` interface changed in the wrong direction - why is it taking a `runParams` instead of just a `plan`? I couldn't see a reason.
It was always bad, but I think it became worse now - at least before it seemed likely to a reader that the second argument (the plan) matters and the first one is random crap. 
</blockquote></details>

so what do you suggest?

---

*[pkg/sql/parallel_stmts.go, line 88 at r30](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305#-L3yl-CNFC21ARnFyVIH:-L4c6e9ZLVocfy6OQGL4:bqnsel4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e140bedd72924c16164d0ffa6a15ee2a3632c5/pkg/sql/parallel_stmts.go#L88)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

should this be `params.p.curPlan.close()` ?
</blockquote></details>

No, that's different. by the time this Close runs, params.p.curPlan may be hosting a different plan already (because of concurrent execution)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21305)*
<!-- Sent from Reviewable.io -->
